pkg/agent/controller/ipseccertificate/ipsec_certificate_controller.go

```diff
@@ -0,0 +1,403 @@
+// Copyright 2022 Antrea Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ipseccertificate
+
+import (
+	"context"
+	"crypto"
+	"crypto/rand"
+	"crypto/rsa"
+	"crypto/sha256"
+	"crypto/x509"
+	"crypto/x509/pkix"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"time"
+
+	certificatesv1 "k8s.io/api/certificates/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/wait"
+	clientset "k8s.io/client-go/kubernetes"
+	certutil "k8s.io/client-go/util/cert"
+	csrutil "k8s.io/client-go/util/certificate/csr"
+	"k8s.io/client-go/util/keyutil"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/klog/v2"
+
+	antreaapis "antrea.io/antrea/pkg/apis"
+	"antrea.io/antrea/pkg/ovs/ovsconfig"
+)
+
+const (
+	controllerName = "AntreaAgentIPsecCertificateController"
+	workerItemKey  = "key"
+	minRetryDelay  = 5 * time.Second
+	maxRetryDelay  = 300 * time.Second
+
+	// the mount path for CA certificate in antrea-ipsec container.
+	// StrongSwan will never reads CA certificates from folders other than `/etc/ipsec.d/cacerts`.
+	// Though StrongSwan will automatically load CA certificates from the folder, we set the ca_path in other_configs
+	// to the correct path for better consistency.
+	ipsecContainerCAMountPath = "/etc/ipsec.d/cacerts/ca.crt"
+)
+
+// certificateWaitTimeout controls the amount of time we wait for certificate
+// approval in one iteration.
+var certificateWaitTimeout = 15 * time.Minute
+var defaultCertificatesPath = "/var/run/openvswitch"
+
+// Controller is responsible for requesting certificates by CertificateSigningRequest and configure them to OVS
+type Controller struct {
+	kubeClient      clientset.Interface
+	ovsBridgeClient ovsconfig.OVSBridgeClient
+	nodeName        string
+	queue           workqueue.RateLimitingInterface
+
+	certificateKeyPair *certificateKeyPair
+
+	// the rotation deadline of current certificateKeyPair
+	rotationDeadline time.Time
+	// caPath and privateKeyPath are initialized with NewIPSecCertificateController
+	// should not be changed once Controller starts.
+	caPath, privateKeyPath string
```

Comment:
I think we need to support CA rotating for both user provided and self-signed eventually. Since it's an alpha feature, maybe we could do it in next release.

Reply:
ok. I will test how to rotate CA seamlessly without restarting Antrea components.